sw/airborne/modules/ctrl/optical_flow_landing.c

```diff
@@ -225,213 +283,394 @@ void vertical_ctrl_module_run(bool in_flight)
+      // deal with (unlikely) fast changes in divergence:
+      float max_div_dt = 0.20;
+      if (fabs(new_divergence - divergence) > max_div_dt) {
+        if (new_divergence < divergence) { new_divergence = divergence - max_div_dt; }
+        else { new_divergence = divergence + max_div_dt; }
       }
-    } else {
-      // USE REAL VISION OUTPUTS:
 
-      if (vision_message_nr != previous_message_nr && dt > 1E-5 && ind_hist > 1) {
-        // TODO: this div_factor depends on the subpixel-factor (automatically adapt?)
-        div_factor = -1.28f; // magic number comprising field of view etc.
-        float new_divergence = (divergence_vision * div_factor) / dt;
+      // low-pass filter the divergence:
+      divergence = divergence * of_landing_ctrl.lp_factor + (new_divergence * (1.0f - of_landing_ctrl.lp_factor));
+      previous_message_nr = vision_message_nr;
+      dt = 0.0f;
 
-        if (fabs(new_divergence - divergence) > 0.20) {
-          if (new_divergence < divergence) { new_divergence = divergence - 0.10f; }
-          else { new_divergence = divergence + 0.10f; }
+    } else {
+      // after re-entering the module, the divergence should be equal to the set point:
+      if (ind_hist <= 1) {
+        divergence = of_landing_ctrl.divergence_setpoint;
+        for (i = 0; i < MAX_COV_WINDOW_SIZE; i++) {
+          thrust_history[i] = 0;
+          divergence_history[i] = 0;
+          dt_history[i] = 0;
         }
-        // low-pass filter the divergence:
-        divergence = divergence * of_landing_ctrl.lp_factor + (new_divergence * (1.0f - of_landing_ctrl.lp_factor));
-        previous_message_nr = vision_message_nr;
+        // TODO: is this correct? Shouldn't dt be incremented?
         dt = 0.0f;
-      } else {
-        // after re-entering the module, the divergence should be equal to the set point:
-        if (ind_hist <= 1) {
-          divergence = of_landing_ctrl.divergence_setpoint;
-          for (i = 0; i < COV_WINDOW_SIZE; i++) {
-            thrust_history[i] = 0;
-            divergence_history[i] = 0;
-          }
-          ind_hist++;
-          dt = 0.0f;
-          int32_t nominal_throttle = of_landing_ctrl.nominal_thrust * MAX_PPRZ;
-          stabilization_cmd[COMMAND_THRUST] = nominal_throttle;
-
-        }
-        // else: do nothing, let dt increment
-        return;
+        int32_t nominal_throttle = of_landing_ctrl.nominal_thrust * MAX_PPRZ;
+        stabilization_cmd[COMMAND_THRUST] = nominal_throttle;
       }
+      // else: do nothing, let dt increment
+      return;
     }
+  }
+
+  if (in_flight) {
 
     /***********
     * CONTROL
     ***********/
 
-    int32_t nominal_throttle = of_landing_ctrl.nominal_thrust * MAX_PPRZ; \
+    float err;
+    int32_t thrust;
 
     // landing indicates whether the drone is already performing a final landing procedure (flare):
     if (!landing) {
 
+      // First seconds, don't do anything crazy:
+      if (module_active_time_sec < 2.5f) {
```

Comment:
If you run the low passing of the divergence in the vertical_ctrl_optical_flow_cb() you will continuously have a low passed divergence estimate and not need to wait.